### PR TITLE
Fixes #8729: Create a rudder agent run --force option

### PR DESCRIPTION
**Pull request by peckpeck**
https://www.rudder-project.org/redmine/issues/8729

### CODE REVIEW

**Comment by amousset** [comment]
This will not work as the agent will stop at the beginning of the run because of the disable-agent file. We can pass a class to force running anyway, but this needs to be done in the techniques.

**Comment by peckpeck** [comment]
Good catch, will add this to techniques too

**Comment by peckpeck** [comment]
Commit modified

File: share/commands/agent-run [hunk 1]
```diff
@@ -99,14 +103,19 @@ while getopts "iIvdqwrRmcb:D:" opt; do
...
 
 printf "${VERSION}\nNode uuid: ${UUID}\n"
 
-if [ -e /opt/rudder/etc/disable-agent ]; then
+if [ ${FORCE} -eq 0 ] && [ -e /opt/rudder/etc/disable-agent ]; then
   printf "\n${RED}error${NORMAL}: The Rudder agent is currently disabled. You can enable it with 'rudder agent enable'.\n"
   exit 1
 fi
 
+[ ${FORCE} -eq 1 ] && CLASS="${CLASS} -D run_disabled"
```
Comment (amousset) [inline share/commands/agent-run]: Multiple `-D` options are only supported starting from CFEngine 3.7.2 (I.E. Rudder 3.3). We should use `-D class1,class2` insted. See https://tracker.mender.io/browse/CFE-1941 for details.

**Comment by peckpeck** [comment]
Commit modified

File: share/commands/agent-run [hunk 2]
```diff
@@ -45,6 +47,8 @@ SUMMARY_ONLY=0
 QUIET=0
 # Display full strings
 FULL_STRINGS=0
+# Ignore agent-disable flag
```
Comment (amousset) [inline share/commands/agent-run]: disable-agent

**Comment by peckpeck** [comment]
Commit modified

File: share/commands/agent-run [hunk 3]
```diff
@@ -45,6 +47,8 @@ SUMMARY_ONLY=0
 QUIET=0
 # Display full strings
 FULL_STRINGS=0
+# Ignore disablei-agent flag
```
Comment (amousset) [inline share/commands/agent-run]: disablei-agent -> disable-agent

**Comment by peckpeck** [comment]
Commit modified

**Comment by jooooooon** [comment]
This looks good to me, aside from the class naming comment in #993

**Comment by jooooooon** [comment]
OK, naming agreed up. This is ready for merge but depends on Normation/rudder-techniques#993

**Comment by peckpeck** [comment]
Do not merge, I will add the force option to rudder agent inventory first

**Comment by jooooooon** [comment]
OK :)

**Comment by peckpeck** [comment]
Commit modified

**Comment by peckpeck** [comment]
ready for review

File: share/commands/agent-inventory [hunk 4]
```diff
@@ -24,5 +24,16 @@
...
+# @man +
+# @man *-f*: run the agent even if it is disabled
+
+OPTS=""
+argcount=$#
+# Process all options in order (getopts doesn't set OPTIND in case of error)
+for OPTIND in `seq 1 ${argcount}`
+do
+  getopts ":f" opt # ignore getopt errors
+  [ "${opt}" = "f" ] && OPTS="-D ignore_disable_inventory"
```
Comment (jooooooon) [inline share/commands/agent-inventory]: There is already a "-D" option in the command call below. As Alexis pointed out, cf-agent doesn't support multiple -D options, we need to use the -D class1,class2 syntax.

File: share/commands/agent-inventory [hunk 5]
```diff
@@ -24,5 +24,16 @@
...
 # @man *-m*: run the agent with multiline output
+# @man +
+# @man *-f*: run the agent even if it is disabled
+
+OPTS=""
+argcount=$#
+# Process all options in order (getopts doesn't set OPTIND in case of error)
+for OPTIND in `seq 1 ${argcount}`
+do
+  getopts ":f" opt # ignore getopt errors
```
Comment (jooooooon) [inline share/commands/agent-inventory]: Shouldn't this just be getopts "f"? I thought the ":" meant "the previous option takes an argument"

Comment (peckpeck) [inline share/commands/agent-inventory]: : at the begining of options means do not print errors

File: share/commands/agent-inventory (hunk 4 above)
Comment (peckpeck) [inline share/commands/agent-inventory]: Thats why there is a code in rudder agent run to merge multiple -D

**Comment by jooooooon** [comment]
OK! :+1:

**Comment by peckpeck** [comment]
OK, merging this PR

**Pull request merged**